REACH Dashboard

lwyszomi opened the pull request:
Hi @emord,

I sent PR with the REACH Dashboard but this is still WIP so please don't merge.

Need QA: yes when PR all changes will be finished
Environment: Don't know
Locally tested: yes

What needs to be finished:
- [x] month/year filter UI and logic in the backend
- [x] location filter UI and logic in the backend
- [x] UI fixes like active tab in the left menu
- [x] Add tests for UI

Regards,
Łukasz

Commit:
WIP - REACH Dahsboard - created module for reports

Commit:
[WIP] - create design for REACH Dashboard

Commit:
[WIP] Program Overview Report

lwyszomi commented:
![program_overview](https://user-images.githubusercontent.com/429541/51896269-5cd01a80-23ac-11e9-8d1e-fd285f5c5d70.png)

lwyszomi commented on settings.py:
```diff
@@ -2094,7 +2095,7 @@
     'icds-cas': 'custom.icds_reports',
     'icds-dashboard-qa': 'custom.icds_reports',
-    'reach-test': 'custom.icds_reports',
+    'reach-test': 'custom.reach',
```
@emord probably we need to add a possibility to enable two custom modules for one domain?

emord replied:
I think that in the future we may need that. Currently they will be used in separate domains so I do not believe it will be necessary now.

Commit:
linter

Commit:
[WIP] REACH Dashboard - Added month/year filter

lwyszomi commented:
![reach-dashboard-month-year-filter](https://user-images.githubusercontent.com/429541/51916769-49d63e00-23de-11e9-9b3c-a9c6b884e0e1.gif)

emord commented on custom/reach/const.py:
```diff
@@ -0,0 +1,23 @@
+from __future__ import unicode_literals
+
+INDICATOR_LIST = {
+    'registered_eligible_couples': 'Registered Eligible Couples',
```
Can you mark the report names and other strings for translation?

emord commented on custom/reach/templates/reach/menus/left_menu.html:
```diff
@@ -0,0 +1,20 @@
+        <ul class="nav nav-list" >
+            <li><a href="{% url 'program_overview' domain %}">Program Overview</a></li>
+            <li><a href="{% url 'unified_beneficiary' domain %}">Unified Beneficiary Report</a></li>
+            <li>report 1</li>
```
For now, report 1 - 5 can be left off the sidebar

emord reviewed (commented):
Thanks for uploading the WIP and screenshots. Looks good in general, I'll do a more thorough review as I get a chance, just some small feedback for now.

It looks like these screenshots are slightly off from the spec. CommCare should be replaced with the dimagi logo and the program overview text should be where REACH Dashboard is.

Also the colors look quite a bit brighter than the spec. I've asked Wouter if there are any specific colors he's been provided.

lwyszomi commented:
@emord Thanks for the feedback, I will add all changes, but Could you send me a Dimagi Logo because I couldn't find any in good quality.

Commit:
change place of title add translations support

Commit:
[WIP] Location filter

Commit:
Added location Filter

emord commented on custom/reach/urls.py:
```diff
@@ -0,0 +1,21 @@
+
+urlpatterns = [
+    url(r'^reach_dashboard/', include(dashboardurls)),
+    url(r'^reach_dashboard_data/', include(dataurls)),
```
Can this be changed to `aaa_dashboard` and `aaa_dashboard/data` ?

lwyszomi replied:
:+1: Done

Commit:
[WIP] Added tests and code clean

lwyszomi commented on custom/reach/templates/reach/dashboard.html:
```diff
@@ -0,0 +1,78 @@
+    {% javascript_libraries underscore=True hq=True %}
+    <script src="{% static 'moment/moment.js' %}"></script>
+
+    <script src="{% static 'select2-3.5.2-legacy/select2.js' %}"></script>
```
:+1: Done

lwyszomi commented on custom/reach/templates/reach/dashboard.html:
```diff
@@ -0,0 +1,78 @@
+<head>
+    <link type="text/css" rel="stylesheet" media="all" href="{% static 'bootstrap/dist/css/bootstrap.css' %}"/>
+    <link type="text/css" rel="stylesheet" media="all" href="{% static 'bootstrap/dist/css/bootstrap.css' %}"/>
+    <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.1/css/bootstrap-combined.min.css"/>
```
I cleaned the code

lwyszomi commented on custom/reach/templates/reach/dashboard.html:
```diff
@@ -0,0 +1,78 @@
+<html>
+<head>
+    <link type="text/css" rel="stylesheet" media="all" href="{% static 'bootstrap/dist/css/bootstrap.css' %}"/>
+    <link type="text/css" rel="stylesheet" media="all" href="{% static 'bootstrap/dist/css/bootstrap.css' %}"/>
```
I cleaned the code

lwyszomi commented:
Hi @emord 

I updated the PR and I added changes requested in the comments, also I added tests for UI but still I need to add tests for the location filter

Please let me know if you have any questions

Regards,
Łukasz

Commit:
linter

emord commented on custom/reach/const.py:
```diff
@@ -0,0 +1,25 @@
+    'aqua': '#05EDFF',
+    'orange': '#FFA500',
+    'darkorange': '#FF8C00',
+}
```
Can you update with the following codes for the correct reports:

Registered Eligible Couples : #725CA4
Registered Pregnancies : #04AEE6
Registered Children : #F5BB5D
Couples using Family Planning Method : #1CC6CC
High Risk Pregnancies : #FF8300
Institutional Deliveries : #004EBC

emord reviewed (commented):
You can use the following logo ![dimagi-logo-rgb](https://user-images.githubusercontent.com/1471773/52233672-a7123980-288d-11e9-994f-ae9b9988c87e.png)

Commit:
Added tests for location filter

Commit:
update logo and colors

lwyszomi commented:
Hi @emord 

I updated the logo and colours, also I finished all tests for UI, so Pr is ready to merge and QA.

Please let me know if you have any requests.

Screenshot after updates:
![dashboard](https://user-images.githubusercontent.com/429541/52277543-43ebda00-2955-11e9-8403-1612e832198c.png)

Regards,
Łukasz

emord commented on custom/reach/reports/reports.py:
```diff
@@ -0,0 +1,21 @@
+@location_safe
+class DashboardReport(CustomProjectReport):
+    slug = 'dashboard_report'
+    name = 'REACH Dashboard'
```
```suggestion
    name = 'AAA Convergence Dashboard'
```

lwyszomi replied:
:+1:

emord commented on custom/reach/reports/reports.py:
```diff
@@ -0,0 +1,21 @@
+    @classmethod
+    def show_in_navigation(cls, domain=None, project=None, user=None):
+        #TODO replace to new feature flag
+        return True
```
```suggestion
        return False
```

emord replied:
Hiding this for now just in case we need to

lwyszomi replied:
:+1:  I added a new feature flag for this, so we can enable/disable reports from UI without new PR.

emord commented on custom/reach/static/reach/css/reach_dashboard.css:
```diff
@@ -0,0 +1,157 @@
+
+.modal-title {
+    font-size:30px;
+color:#004ebc;
```
```suggestion
    color:#004ebc;
```

lwyszomi replied:
:+1:

emord commented on custom/reach/views.py:
```diff
@@ -0,0 +1,142 @@
+
+    def get_context_data(self, **kwargs):
+        kwargs['domain'] = self.domain
+        # TODO add logic for user role type possible options should be MoHFW or MWCD
```
When is the logic going to be added? For now I think we can decide that it will be in the user's custom data in a field called `ministry`

emord replied:
Can you add a separate dropdown in this filter that allows web users to select between the hierarchies? This dropdown should be hidden fro mobile workers

emord commented on custom/reach/const.py:
```diff
@@ -0,0 +1,25 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from django.utils.translation import ugettext_lazy as _
```
I think these need to be `ugettext_noop` as they are being evaluated before a language is knownw

lwyszomi replied:
:+1:

emord commented on custom/reach/views.py:
```diff
@@ -0,0 +1,142 @@
+                    'format': NUMERIC,
+                    'color': COLORS['violet'],
+                    'numerator': 71682,
+                    'denominator': 140098,
```
Why does this need a `denominator`? This would not be shown as its not a percent

emord commented on custom/reach/views.py:
```diff
@@ -0,0 +1,142 @@
+                    'numerator': 71682,
+                    'denominator': 140098,
+                    'past_month_numerator': 69354,
+                    'past_month_denominator': 130098
```
Instead of `past_month_numerator` and `past_month_denominator`, I believe that this should just be `past_month_value` since there is no numerator or denominator

lwyszomi replied:
:+1:

emord commented on custom/reach/templates/reach/reports/unified_beneficiary.html:
```diff
@@ -0,0 +1,11 @@
+
+{% block js-inline %}
+    {{ block.super }}
+{% endblock %}
```
Is this `block` necessary since it only calls super?

lwyszomi replied:
I added this for future to remember that when we will be working on the unified beneficiary report UI we will only have to add the import of js file, something like example template for future reports

emord commented on custom/reach/templates/reach/dashboard.html:
```diff
@@ -0,0 +1,66 @@
+</div>
+
+{% block js-inline %}
+    <script src="{% static 'reach/js/dashboard.js' %}"></script>
```
Why isn't this js file with the others in <head> ?

emord commented on custom/reach/templates/reach/reports/program_overview.html:
```diff
@@ -0,0 +1,51 @@
+            $('#reach-dashboard').koApplyBindings(programOverviewModel);
+            programOverviewModel.callback();
+        })
+    </script>
```
Can this be put into a separate json file? Why can't it be included in `program_overview.js`?

Also why is this in `js-inline` instead of `js`

orangejenny replied:
Confirming this should be in a separate js file.

lwyszomi replied:
I moved this to the ```program_overview.js``` file

emord commented on custom/reach/static/reach/js/filters/month_year_filter.js:
```diff
@@ -0,0 +1,68 @@
+
+            updateMonths(self.selectedYear());
+
+            for (var year=2017; year <= moment().year(); year++ ) {
```
We should start this at 2019

lwyszomi replied:
:+1:

emord reviewed (commented):
sorry for the questions on the javascript stuff. It's been a while since I wrote a new page in javascript, so they're mostly for my own information

Commit:
requested suggestions

Commit:
Merge branch 'master' into lw/reach-dashboard

emord commented on custom/reach/static/reach/js/reports/program_overview.js:
```diff
@@ -0,0 +1,90 @@
+            return value.toFixed(2) + " %";
+        };
+
+        self.diffBetweenMonths = function () {
```
I'm not sure that this is the correct way to compare previous month and current month numbers but i will check

emord reviewed (approved):
I think this is fine to merge as is. Followup commits can be done in a separate PR
